keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+            * [Summary](#summary-1)
+         * [Existing Problems](#existing-problems)
+            * [Non-generic Capacity-based Garbage Collection](#non-generic-capacity-based-garbage-collection)
+            * [Non-generic TTL-based Garbage Collection](#non-generic-ttl-based-garbage-collection)
+            * ["Root" Resources Without an Owner](#root-resources-without-an-owner)
+            * [Immutable Pattern: Short-lived, Frequently-created Resources](#immutable-pattern-short-lived-frequently-created-resources)
+            * [Cannot Centralize Cleanup of Multiple Resources with OwnerReferences](#cannot-centralize-cleanup-of-multiple-resources-with-ownerreferences)
+      * [Proposal](#proposal)
+         * [Generic TTL-based Garbage Collection](#generic-ttl-based-garbage-collection)
+         * [Clean Up Resources After Use](#clean-up-resources-after-use)
+         * [Summary](#summary-2)
+         * [User Stories](#user-stories)
+            * [Story 1: Clean up old Jobs](#story-1-clean-up-old-jobs)
+            * [Story 2: Secrets and ConfigMaps](#story-2-secrets-and-configmaps)
+            * [Story 3: CustomResourceDefinitions (CRDs) and custom resources](#story-3-customresourcedefinitions-crds-and-custom-resources)
+         * [Implementation Details](#implementation-details)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Future Work](#future-work)
+         * [Generic Capacity-based Garbage Collection](#generic-capacity-based-garbage-collection)
+         * [Namespace-wide Garbage Collection Policy](#namespace-wide-garbage-collection-policy)
+         * [Move Event TTL out of etcd TTL](#move-event-ttl-out-of-etcd-ttl)
+         * [Make Pod GC and Event GC Work Better with Garbage Collector](#make-pod-gc-and-event-gc-work-better-with-garbage-collector)
+         * [Move deletionGracePeriodSeconds from metadata to Pod spec](#move-deletiongraceperiodseconds-from-metadata-to-pod-spec)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+We propose to expand Kubernetes garbage collection to handle common uses which
+are awkward today. We will add an object-lifecycle-respecting TTL mechanism to
+the garbage collector. We will also allow users to mark resource objects to be
+cleaned up by the garbage collector after those objects are no longer in use:
+either terminated, completed, or not referenced by other resource objects.
+
+## Motivation
+
+Make it easy for users to clean up frequently-created, short-lived Kubernetes
+resources objects.
+
+One common use case is [ConfigMaps][] and [Secrets][]. The users roll out
+ConfigMaps and Secrets by creating new ConfigMaps or Secrets and updating the
+references to those ConfigMaps or Secrets in containers ([#22368][]). The old
+ConfigMaps and Secrets, if not referenced by any container, are no longer
+needed. 
```

Comment:
ConfigMaps and secrets can be consumed by other means than containers right? You can have an operator that is reading the content of an configMap or secret (can be a CA or something)...

Reply:
(also you can reference configMap via downward api)

Reply:
Resource objects, including ConfigMaps/Secrets, can decide whether they want to use this feature (GC after unused) or not. 

Referencing via downward api is also covered by this proposal. The schema information can be encoded in OpenAPI spec. If the operator uses an API object to read/reference a ConfigMap/Secret, the usage can be discovered in the OpenAPI spec. If the operator simply reads/gets the ConfigMap/Secret, I would imagine that such ConfigMap/Secret won't be massively created or short-lived but be updated directly. In this case, they don't need GC after unused.

Reply:
@janetkuo Just to verify, it's opt-in to non-referenced ConfigMaps/Secrets being GC'd if not attached to a container rather than opt-out. Right?